source/common/interruptDispatcher.s

```diff
@@ -67,46 +74,65 @@ BEGIN_ASM_FUNC IntrMain
     orr     r3, r3, r1
     str     r3, [r0]
 #endif
-endsetflags:
 
+    // find the highest set IRQ bit
+    // input:
+    // r1 = interrupt mask
+    // r2 = irq table address
+    // output:
+    // r1 = target interrupt mask
+    // r2 = target irq table address
+    // r0, r3 can be used freely
 findIRQ:
-
-    ldr     r0, [r2, #4]
-    cmp     r0, #0
+#ifdef ARM9
+    clz     r0, r1
+    eor     r0, r0, #31
+#else
+    mov     r0, #0
+    cmp     r1, #0x10000
+    movcs   r1, r1, lsr #16
+    addcs   r0, r0, #16
+    tst     r1, #0xFF00
+    movne   r1, r1, lsr #8
+    addne   r0, r0, #8
+    tst     r1, #0xF0
+    movne   r1, r1, lsr #4
+    addne   r0, r0, #4
+    tst     r1, #0xC
+    movne   r1, r1, lsr #2
+    addne   r0, r0, #2
+    // r1 is now equal to 0, 1, 2 or 3
+    // if it's equal to 2 or 3, add 1 to r0
+    add     r0, r0, r1, lsr #1
+#endif
+    mov     r1, #1
+    mov     r1, r1, lsl r0
+    add     r2, r2, r0, lsl #2
+
+    // clear the IRQ now being serviced
+    str     r1, [r12, #-4]
+
+    // compare dummy IRQ address with found IRQ address
+    // this skips some setup required for jumping to an IRQ handler
+    ldr     r0, =irqDummy
+    ldr     r1, [r2]
+    cmp     r1, r0
     beq     no_handler
-    ands    r0, r0, r1
-    bne     jump_intr
-    add     r2, r2, #8
-    b       findIRQ
-
-no_handler:
-
-    str     r1, [r12, #4]           // IF Clear
-    ldmfd   sp!, {r0-r1, r12, lr}   // {spsr, IME, REG_BASE, lr_irq}
-    str     r1, [r12, #0x208]       // restore REG_IME
-    mov     pc, lr
-
-jump_intr:
-
-    ldr     r1, [r2]                // user IRQ handler address
-    cmp     r1, #0
-    bne     got_handler
-    mov     r1, r0
-    b       no_handler
 
 got_handler:
 
-    str     r0, [r12, #4]           // IF Clear
-
     mrs     r2, cpsr
-    mov     r3, r2
-    bic     r3, r3, #0xdf           // Enable IRQ & FIQ. Set CPU mode to System
+    bic     r3, r2, #0xdf           // Enable IRQ & FIQ. Set CPU mode to System
```

Comment:
Can you use the `CPSR` defines in `include/nds/cpu_asm.h` instead of having this number here? I guess you could also replace the magic number below, that's up to you.